docs/articles/documentation/guides/basic-guides/best-practices.md

```diff
@@ -0,0 +1,333 @@
+---
+layout: docs
+title: Best Practices
+permalink: /documentation/guides/basic-guides/best-practices.html
+---
+# Best Practices
+
+This article describes the recommended ways to test with TestCafe and covers the following topics:
+
+* [E2E Tests Scope](#e2e-tests-scope)
+* [Smart Assertions](#smart-assertions)
+* [Use of Page Objects](#use-of-page-objects)
+* [Use of Roles for Login](#use-of-roles-for-login)
+* [File structure](#file-structure)
+* [Setup And Teardown](#setup-and-teardown)
+* [Selectors strategy](#selectors-strategy)
+
+## E2E Tests Scope
+
+During the functional (also known as 'end-to-end') testing the application is tested from beginning to end. This is in contrast with unit or integration tests, which focus on a specific part (or integration between parts) of the application.
+
+TestCafe is a tool built for functional testing. Do not use it to perform non-functional testing (like performance or load testing). Such tests would not yield any conclusive results.
+
+In your end-to-end tests, try to replicate real user's actions. Don't test exceptions - these are better tested with unit and integration tests. Test the general business logic of your application and abstain from rare use cases and edge scenarios.
+
+Write less E2E tests. End-to-end tests are slow by nature, their amount should be drastically lower than that of unit or integration tests.
+
+## Smart Assertions
+
+In end-to-end web testing, unpredictable factors (like network lag, processor, or memory bottlenecks in containers) can interfere with the assertions and produce inconsistent test results. Such tests are inconclusive (sometimes called 'flaky').
+
+TestCafe includes a [Smart Assertion Query Mechanism](../../guides/basic-guides/assert.md#smart-assertion-query-mechanism). This mechanism introduces wait time for all the assertions: if an assertion fails, it retries multiple times within a timeout. That reduces random impact and stabilizes the tests without performance trade-offs.
```

Comment:
Note that a colon is used when a list follows. In this case, it's better to replace it with a semicolon, or split the sentences into two. For example:

`This mechanism introduces a wait time for all assertions; if an assertion fails, it retries multiple times within a timeout.`

or

`This mechanism introduces a wait time for all assertions. If an assertion fails, it retries multiple times within a timeout.`